llvm/include/llvm/Analysis/TargetTransformInfo.h

```diff
@@ -1392,6 +1392,19 @@ class TargetTransformInfo {
                                      unsigned Index = -1, Value *Op0 = nullptr,
                                      Value *Op1 = nullptr) const;
 
+  /// \return The expected cost of vector Insert and Extract.
+  /// Use -1 to indicate that there is no information on the index value.
+  /// This is used when the instruction is not available; a typical use
+  /// case is to provision the cost of vectorization/scalarization in
+  /// vectorizer passes.
+  InstructionCost getVectorInstrCost(
+      unsigned Opcode, Type *Val, TTI::TargetCostKind CostKind, unsigned Index,
+      Value *Op0, Value *Op1, Value *Scalar,
+      const DenseMap<std::pair<Value *, unsigned>, SmallVector<Value *, 4>>
```

Comment:
In general in user interfaces I thought we typically avoided passing in `SmallVector<Type, 4>` in favour of `SmallVectorImpl<Type>` so that the user is not tied to a particular initial size?

Reply:
Using SmallVectorImpl<Type> inside a Densemap gives me compile time error. Perhaps thats not supported. But I will try to simplify the data structure, as you have suggested in the other comment, to avoid data  duplication.